extmod/modbluetooth: Remove non-sync event support.

jimmo opened the pull request:
This was the original implementation before we had a way to support running
events synchronously from the BLE stack.

ESP32 was the last remaining port using this, and now that it's been
updated (in https://github.com/micropython/micropython/commit/5dbb822ca4a809ac5cb4513afb0411b4eb8dc3cf & https://github.com/micropython/micropython/commit/e05d0a6335be6a45f29f5d987091b858ad378cfa) we can remove this.

Removes the MICROPY_PY_BLUETOOTH_ENTER. This was previously a no-op on
ports (e.g. stm32) using sync events anyway. Replaced with
MICROPY_BEGIN_ATOMIC_SECTION where necessary (e.g. unix mpbthciport.c).

Also makes pairing&bonding enabled by default (this was effectively the
current behavior as it was enabled when sync events were enabled).

jimmo committed:
extmod/modbluetooth: Remove non-sync event support (via ringbuf).

This ringbuffer-based approach was the original implementation before we
had a way to support running events synchronously from the BLE stack.

Zephyr is the last remaining port using this. ESP32 no longer requires it
as of 5dbb822ca & e05d0a633. The Zephyr BLE support is about to be updated
anyway.

This commit also removes the MICROPY_PY_BLUETOOTH_ENTER/EXIT guards. This
was previously a no-op on ports using sync events (e.g. stm32). Replaced
with MICROPY_BEGIN_ATOMIC_SECTION where necessary it was used in other
locations (e.g. unix mpbthciport.c).

Also makes pairing & bonding enabled by default (this was effectively the
current behavior as it was enabled when sync events were enabled).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge commented:
I'm not sure the zephyr port will work with synchronous events... need to check it.

jimmo committed:
py/mpthread: Add RTOS-support function to run on mp thread.

This abstracts the implementation currently used by modbluetooth.c
to get BLE events from the host stack running in an RTOS task into
a MicroPython task.

This existing implementation only works when the GIL is enabled, so also
add a scheduler-based implementation for non-GIL ports (e.g. Unix).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Use mp_thread_run_on_mp_thread.

This replaces the bluetooth-specific implementation previously enabled when
MICROPY_PY_BLUETOOTH_USE_SYNC_EVENTS_WITH_INTERLOCK was enabled to
synchronously run a host-stack callback on a MicroPython thread.

When MICROPY_PY_THREAD_RTOS is not enabled, this is a no-op and the
call goes directly (e.g. stm32, where the host stack is running in the
scheduler).

Also enables this for esp32 by removing
MICROPY_PY_BLUETOOTH_USE_SYNC_EVENTS_WITH_INTERLOCK and adding
MICROPY_PY_THREAD_RTOS.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
unix/mpconfigport: Allow running with the GIL enabled.

MICROPY_EVENT_POLL_HOOK should bounce the GIL like other ports.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
unix/mpbthciport: Use mp_thread_run_on_mp_thread.

This replaces the previous code that used the scheduler (via thread) to
run the BLE host stack.

Now the host stack can run directly on the thread, and then
modbluetooth.c will use mp_thread_run_on_mp_thread to run the
Python-level callbacks on a MicroPython thread as necessary.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
I have updated this PR to make a common mechanism for having RTOS tasks call into a MicroPython thread.

This replaces the `MICROPY_PY_BLUETOOTH_USE_SYNC_EVENTS_WITH_INTERLOCK` that was previously used specifically for ESP32, and allows the Unix (and soon, Zephyr) port to use the same functionality (and remove all the special scheduler handling in `unix/mpbthciport.c`).

jimmo committed:
extmod/nimble/nimble/nimble_npl_os: Implement synchronisation.

This provides a full implementation of the NimBLE mutex, critical section,
and semaphore. On ports without threading, this is a no-op change, but this
is necessary to make NimBLE's locking functionality work on threaded ports.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/mpthread: Add support for a recursive mutex.

Also provide a no-op mutex when threading is disabled.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth: Improve event ordering handling.

For _IRQ_GATTC_READ_DONE, we should wait for it to ensure that we receive
it before we sent the write (this prevents an ordering race with the
"gattc_write" vs the "_IRQ_GATTC_READ_DONE" line).

For _IRQ_GATTC_WRITE_DONE, don't print it out because we wait for it anyway
(so it will be confirmed to have been generated), and it can race with
the concurrent notify or indicate event.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/mpthread: Add a semaphore primitive.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble: Update to NimBLE 1.6.0.

Implement a UART transport:
The way transports worked changed slightly in NimBLE 1.6.0 anyway, but
this allows us to hook the "send payload" directly, rather than having
to go byte-at-a-time via the tx callback. Also implemenets a mutex on
transmission as nothing prevents NimBLE sending both a CMD and ACL
payload at the same time.

Common implementation of mpnimbleport.c for scheduler-based ports:
Makes stm32, renesas-ra, mimxrt, rp2 use a shared implementation.

Simplify the NPL event queue:
We only need to process the single default queue, so remove the
additional logic to handle the linked list of queues. Also rename the
functions to execute the queues and run timers.

Unix:
Implement the HCI layer more like an RTOS port, where you have a task
running the host stack, and UART data is sent directly to NimBLE from
a UART IRQ. This makes it behave more like ESP32 or Zephyr, where we
only run the Python-level callbacks in a MicroPython thread.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/hci_trace_to_pcap.py: Support colour input.

Filter out escape sequences to set the color. Also allow the "TX:" and
"RX:" used in some places.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dmazzella commented:
Are there any plans for the merge of this PR?

jimmo commented:
> Are there any plans for the merge of this PR?

Yes. See https://github.com/micropython/micropython/pull/8945#issuecomment-1718852697 -- still some more work to go on this, and need to update the Zephyr port to use it.

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.
